src/java/org/apache/cassandra/db/SinglePartitionReadCommand.java

```diff
@@ -1042,15 +1042,28 @@ private boolean canRemoveRow(Row row, Columns requestedColumns, long sstableTime
         // We can remove a row if it has data that is more recent that the next sstable to consider for the data that the query
         // cares about. And the data we care about is 1) the row timestamp (since every query cares if the row exists or not)
         // and 2) the requested columns.
-        if (row.primaryKeyLivenessInfo().isEmpty() || row.primaryKeyLivenessInfo().timestamp() <= sstableTimestamp)
+        
+        boolean isLivenessInfoEmpty = row.primaryKeyLivenessInfo().isEmpty();
+
+        // Compact tables have no concept of primary key liveness info, so this will never be true for them: 
+        if (!isLivenessInfoEmpty && row.primaryKeyLivenessInfo().timestamp() <= sstableTimestamp)
+        {
             return false;
+        }
+        
+        boolean isCQL = metadata().isCQLTable();
 
         for (ColumnDefinition column : requestedColumns)
         {
             Cell cell = row.getCell(column);
-            if (cell == null || cell.timestamp() <= sstableTimestamp)
+            
+            // Non-compact tables with tombstones may need to look at further SSTables to find a live primary key:
+            if (cell == null || cell.timestamp() <= sstableTimestamp || (isCQL && isLivenessInfoEmpty && cell.isTombstone()))
```

Comment:
TODO: This may not be optimal for UPDATE. (i.e. We might still be able to skip the SSTable as long as there is at least some non-tombstone cell.)